source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,263 @@
+#if defined(ENVOY_ENABLE_FULL_PROTOS)
+#include "source/common/protobuf/deterministic_hash.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/hash.h"
+
+namespace Envoy {
+namespace DeterministicProtoHash {
+namespace {
+
+#define REFLECTION_FOR_EACH(get_type, value_type)                                                  \
+  if (field->is_repeated()) {                                                                      \
+    for (const value_type& q : reflection->GetRepeatedFieldRef<value_type>(message, field)) {      \
+      seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)},   \
```

Comment:
wait, why is this safe? can the thing that's being repeated have a pointer in it and thus not really be deterministic in the right way? Shouldn't we try to deep-hash the object?

Reply:
Same thing, the input is only scalars. Now made explicit with a template.